polar-core/src/inverter.rs

```diff
@@ -0,0 +1,203 @@
+//! A Runnable that runs a query and inverts the results in three ways:
+//!
+//! 1. If no results are emitted (indicating failure), return true.
+//! 2. If at least one result is emitted containing a partial, invert the partial's constraints,
+//!    pass the inverted partials back to the parent Runnable via a shared BindingStack, and return
+//!    true.
+//! 3. In all other cases, return false.
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::rc::Rc;
+
+use crate::counter::Counter;
+use crate::error::PolarResult;
+use crate::events::QueryEvent;
+use crate::folder::{fold_operation, Folder};
+use crate::kb::Bindings;
+use crate::partial::Constraints;
+use crate::runnable::Runnable;
+use crate::terms::{Operation, Operator, Term, Value};
+use crate::vm::{Binding, BindingStack, Goals, PolarVirtualMachine};
+
+#[derive(Clone)]
+pub struct Inverter {
+    vm: PolarVirtualMachine,
+    bindings: Rc<RefCell<BindingStack>>,
+    bsp: usize,
+    results: Vec<BindingStack>,
+}
+
+impl Inverter {
+    pub fn new(
+        vm: &PolarVirtualMachine,
+        goals: Goals,
```

Comment:
The calling VM decides which goals to initialize an Inverter with. If the Inverter were initialized with more than a single Query goal, it would invert any results emitted in the evaluation of all goals.